src/Agent.Listener/Agent.cs

```diff
@@ -164,48 +164,58 @@ void CtrlCHandler(object sender, EventArgs e)
+                    bool skipMessageDeletion = false;
+                    try
                     {
-                        try
+                        skipMessageDeletion = false;
+                        message = await listener.GetNextMessageAsync(token); //get next message
+
+                        if (string.Equals(message.MessageType, AgentRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase))
                         {
-                            message = await listener.GetNextMessageAsync(token); //get next message
-
-                            if (string.Equals(message.MessageType, AgentRefreshMessage.MessageType, StringComparison.OrdinalIgnoreCase))
-                            {
-                                Trace.Warning("Referesh message received, but not yet handled by agent implementation.");
-                            }
-                            else if (string.Equals(message.MessageType, JobRequestMessage.MessageType, StringComparison.OrdinalIgnoreCase))
-                            {
-                                var newJobMessage = JsonUtility.FromString<JobRequestMessage>(message.Body);
-                                workerManager.Run(newJobMessage);
-                            }
-                            else if (string.Equals(message.MessageType, JobCancelMessage.MessageType, StringComparison.OrdinalIgnoreCase))
-                            {
-                                var cancelJobMessage = JsonUtility.FromString<JobCancelMessage>(message.Body);
-                                workerManager.Cancel(cancelJobMessage);
-                            }
+                            Trace.Warning("Referesh message received, but not yet handled by agent implementation.");
                         }
-                        finally
+                        else if (string.Equals(message.MessageType, JobRequestMessage.MessageType, StringComparison.OrdinalIgnoreCase))
                         {
-                            if (message != null)
-                            {
-                                //TODO: make sure we don't mask more important exception
-                                await DeleteMessageAsync(message);
-                                message = null;
-                            }
+                            var newJobMessage = JsonUtility.FromString<JobRequestMessage>(message.Body);
+                            jobDispatcher.Run(newJobMessage);
+                        }
+                        else if (string.Equals(message.MessageType, JobCancelMessage.MessageType, StringComparison.OrdinalIgnoreCase))
+                        {
+                            var cancelJobMessage = JsonUtility.FromString<JobCancelMessage>(message.Body);
+                            bool findJobToCancel = jobDispatcher.Cancel(cancelJobMessage);
+                            skipMessageDeletion = autoUpdateInProgress && !findJobToCancel;
```

Comment:
Maybe rename the variable to jobCancelled or jobFound. findJobToCancel sounds like the caller now needs to go find the job. :-)